### PR TITLE
lte: Add usrsock daemon for ALT1250 LTE modem

### PR DESCRIPTION
**Pull request by SPRESENSE**
## Summary
Add usrsock daemon for ALT1250 LTE modem.
This pull request depends on https://github.com/apache/nuttx/pull/9463. Therefore, please merge https://github.com/apache/nuttx/pull/9463 first.

## Impact
alt1250 daemon

## Testing
alt1250 driver

### CODE REVIEW

**Review by jerpelea — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

File: include/lte/lapi.h [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+ ****************************************************************************/
+
+int lapi_evtyield(int timeout_ms);
+
+/****************************************************************************
+ * Name: lapi_req
+ ****************************************************************************/
+
+int lapi_req(uint32_t cmdid, FAR void *inp, size_t insz, FAR void *outp,
+  size_t outsz, FAR void *cb);
```
Comment (pkarashchenko) [inline include/lte/lapi.h:74]:
```suggestion
int lapi_req(uint32_t cmdid, FAR void *inp, size_t insz, FAR void *outp,
             size_t outsz, FAR void *cb);
```

File: include/lte/lte_api.h [hunk 2]
```diff
@@ -0,0 +1,1579 @@
...
+ *
+ * [in] callback: Callback function to notify that
+ *                change of PIN is completed.
+ *
+ * On success, 0 is returned. On failure,
+ * negative value is returned according to <errno.h>.
+ */
+
+int lte_change_pin(int8_t target_pin, char *pincode,
+                   char *new_pincode, change_pin_cb_t callback);
```
Comment (pkarashchenko) [inline include/lte/lte_api.h:848]: FAR

File: include/lte/lte_api.h [hunk 3]
```diff
@@ -0,0 +1,1579 @@
...
+ *
+ * [out] attemptsleft: Number of attempts left.
+ *                     Set only if failed.
+ *
+ * On success, 0 is returned. On failure,
+ * negative value is returned according to <errno.h>.
+ */
+
+int lte_change_pin_sync(int8_t target_pin, char *pincode,
+                        char *new_pincode, uint8_t *attemptsleft);
```
Comment (pkarashchenko) [inline include/lte/lte_api.h:817]: FAR

File: include/lte/lte_fwupdate.h [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __APPS_INCLUDE_LTE_LTE_FW_API_H
```
Comment (pkarashchenko) [inline include/lte/lte_fwupdate.h:21]:
```suggestion
#ifndef __APPS_INCLUDE_LTE_LTE_FWUPDATE_H
```

File: include/lte/lte_fwupdate.h [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+ *
+ * - LTEFW_RESULT_NOT_ENOUGH_INJECTSTORAGE
+ * - LTEFW_RESULT_DELTAIMAGE_HDR_CRC_ERROR
+ * - LTEFW_RESULT_DELTAIMAGE_HDR_UNSUPPORTED
+ * - EINVAL
+ * - ENODATA
+ *
+ */
+
+int ltefwupdate_injectrest(const char *rest_data, int len);
```
Comment (pkarashchenko) [inline include/lte/lte_fwupdate.h:101]: FAR

File: include/lte/lte_log.h [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+ * Return value : Returns 0 on success. On error, a negative value is
+ *                returned. The following values may be returned
+ *                in error cases.
+ *
+ * - ENOBUFS
+ * - ENOTSUP
+ *
+ */
+
+int lte_log_collect(char output_fname[], size_t len);
```
Comment (pkarashchenko) [inline include/lte/lte_log.h:72]: Maybe can go with more traditional `char *output_fname`. I'm not sure if C89 will allow such constructions

File: include/lte/lte_log.h [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+ * Return value : Returns the 0 on success.
+ *                On error, a negative value is returned. The following
+ *                values may be returned in error cases.
+ *
+ * - ENAMETOOLONG
+ * - ENOTSUP
+ *
+ */
+
+int lte_log_remove(const char *filename);
```
Comment (pkarashchenko) [inline include/lte/lte_log.h:157]: FAR

File: include/lte/lte_lwm2m.h [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+
+typedef void (*lwm2mstub_exec_cb_t)(int seq_no, int srv_id,
+              struct lwm2mstub_instance_s *inst);
+
+typedef void (*lwm2mstub_ovstart_cb_t)(int seq_no, int srv_id,
+              struct lwm2mstub_instance_s *inst, char *token,
+              struct lwm2mstub_ovcondition_s *cond);
+
+typedef void (*lwm2mstub_ovstop_cb_t)(int seq_no, int srv_id,
+              struct lwm2mstub_instance_s *inst, char *token);
```
Comment (pkarashchenko) [inline include/lte/lte_lwm2m.h:250]: FAR

File: lte/alt1250/alt1250_atcmd.c [hunk 9]
```diff
@@ -0,0 +1,932 @@
...
+{
+  atcmd_postproc_t proc;
+  unsigned long arg;
+};
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static void *atcmd_oargs[3];
```
Comment (pkarashchenko) [inline lte/alt1250/alt1250_atcmd.c:60]: FAR

File: lte/alt1250/alt1250_atcmd.c [hunk 10]
```diff
@@ -0,0 +1,932 @@
...
+static int postproc_internal_atcmd(FAR struct alt1250_s *dev,
+                                   FAR struct alt_container_s *reply,
+                                   FAR struct usock_s *usock,
+                                   FAR int32_t *usock_result,
+                                   FAR uint32_t *usock_xid,
+                                   FAR struct usock_ackinfo_s *ackinfo,
+                                   unsigned long arg)
+{
+  int ret = REP_NO_ACK;
+  struct atcmd_postprocarg_t *parg = (struct atcmd_postprocarg_t *)arg;
```
Comment (pkarashchenko) [inline lte/alt1250/alt1250_atcmd.c:81]: FAR

File: lte/alt1250/alt1250_atcmd.c [hunk 11]
```diff
@@ -0,0 +1,932 @@
...
+
+  dev->recvfrom_processing = false;
+
+  err_alt1250("Internal ATCMD Resp : %s\n", (char *)reply->outparam[0]);
+
+  if (parg->proc != NULL)
+    {
+      ret = parg->proc(dev, reply,
+        (FAR char *)reply->outparam[0], *(int *)reply->outparam[2],
```
Comment (pkarashchenko) [inline lte/alt1250/alt1250_atcmd.c:90]:
```suggestion
        (FAR char *)reply->outparam[0], *(FAR int *)reply->outparam[2],
```

File: lte/alt1250/alt1250_dbg.h [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+#include <nuttx/config.h>
+#include <stdio.h>
+#include <debug.h>
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#ifdef CONFIG_LTE_ALT1250_DEBUG_MSG
+# define err_alt1250(v, ...) _err(v, ##__VA_ARGS__)
```
Comment (pkarashchenko) [inline lte/alt1250/alt1250_dbg.h:37]:
```suggestion
# define err_alt1250(v, ...) nerr(v, ##__VA_ARGS__)
```

File: lte/alt1250/alt1250_socket.c [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+/****************************************************************************
+ * name: usocket_smssock_readready
+ ****************************************************************************/
+
+void usocket_smssock_readready(FAR struct alt1250_s *dev)
+{
+  int i;
+  FAR struct usock_s *sock;
+
+  for (i = 0; i < ARRAY_SZ(dev->sockets); i++)
```
Comment (pkarashchenko) [inline lte/alt1250/alt1250_socket.c:168]:
```suggestion
  for (i = 0; i < nitems(dev->sockets); i++)
```

File: lte/alt1250/alt1250_socket.h [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __LTE_ALT1250_ALT1250_SOCKET_H__
```
Comment (pkarashchenko) [inline lte/alt1250/alt1250_socket.h:21]:
```suggestion
#ifndef __APPS_LTE_ALT1250_ALT1250_SOCKET_H
```

File: lte/alt1250/alt1250_socket.h [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+
+#include "alt1250_usockif.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define SELECT_WRITABLE (1 << 1)
+#define SELECT_READABLE  (1 << 2)
+#define SELECTABLE_MASK  (SELECT_WRITABLE | SELECT_READABLE)
```
Comment (pkarashchenko) [inline lte/alt1250/alt1250_socket.h:40]:
```suggestion
#define SELECT_READABLE (1 << 2)
#define SELECTABLE_MASK (SELECT_WRITABLE | SELECT_READABLE)
```

File: lte/alt1250/alt1250_socket.h [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+
+#define USOCKET_SET_REQUEST(sock, _reqid, _xid) { \
+  (sock)->request.reqid = _reqid; \
+  (sock)->request.xid = _xid; \
```
Comment (pkarashchenko) [inline lte/alt1250/alt1250_socket.h:44]:
```suggestion
  (sock)->request.reqid = (_reqid); \
  (sock)->request.xid = (_xid); \
```

File: lte/alt1250/alt1250_socket.h [hunk 17]
```diff
@@ -0,0 +1,249 @@
...
+      if (iii < _OUTPUT_ARG_MAX) \
+        { \
+          (sock)->sock_reply.outparams[iii] = (p); \
+        } \
+      else \
+        { \
+          ASSERT(0); \
+        } \
+    } \
+  while(0);
```
Comment (pkarashchenko) [inline lte/alt1250/alt1250_socket.h:111]:
```suggestion
  while(0)
```

File: lte/alt1250/alt1250_socket.h [hunk 18]
```diff
@@ -0,0 +1,249 @@
...
+#define IS_STATE_READABLE(s) ((s)->select_condition & SELECT_READABLE)
+#define IS_STATE_WRITABLE(s) ((s)->select_condition & SELECT_WRITABLE)
+
+#define IS_SMS_SOCKET(s) ((s)->type == SOCK_SMS)
+
+#define _OUTPUT_ARG_MAX 7
+#define _OPTVAL_LEN_MAX 16
+
+#define usocket_smssock_writeready(d, s) \
+  (usockif_sendtxready((d)->usockfd, USOCKET_USOCKID((s))))
```
Comment (pkarashchenko) [inline lte/alt1250/alt1250_socket.h:137]:
```suggestion
  (usockif_sendtxready((d)->usockfd, USOCKET_USOCKID(s)))
```

File: lte/alt1250/alt1250_usockevent.h [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __LTE_ALT1250_ALT1250_USOCKEVENT_H__
```
Comment (pkarashchenko) [inline lte/alt1250/alt1250_usockevent.h:21]:
```suggestion
#ifndef __APPS_LTE_ALT1250_ALT1250_USOCKEVENT_H
```

File: lte/alt1250/alt1250_usockevent.h [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+
+#include "alt1250_usockif.h"
+#include "alt1250_daemon.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define NEED_CONTAINER_FREE (1<<16)
+#define W_CONTAINER_FREE(a) ((a) | NEED_CONTAINER_FREE)
```
Comment (pkarashchenko) [inline lte/alt1250/alt1250_usockevent.h:38]:
```suggestion
#define NEED_CONTAINER_FREE  (1 << 16)
#define W_CONTAINER_FREE(a)  ((a) | NEED_CONTAINER_FREE)
```

File: lte/alt1250/alt1250_usockevent.h [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+                                    FAR int32_t *usock_result,
+                                    FAR uint32_t *usock_xid,
+                                    FAR struct usock_ackinfo_s *ackinfo);
+
+/****************************************************************************
+ * Public Function Prototypes
+ ****************************************************************************/
+
+int usock_reply(int ufd, int action_code, int32_t result,
+    uint32_t xid, FAR struct usock_ackinfo_s *ackinfo);
```
Comment (pkarashchenko) [inline lte/alt1250/alt1250_usockevent.h:70]:
```suggestion
int usock_reply(int ufd, int action_code, int32_t result,
                uint32_t xid, FAR struct usock_ackinfo_s *ackinfo);
```

File: lte/alt1250/alt1250_util.h [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __LTE_ALT1250_ALT1250_UTIL_H__
```
Comment (pkarashchenko) [inline lte/alt1250/alt1250_util.h:21]:
```suggestion
#ifndef __APPS_LTE_ALT1250_ALT1250_UTIL_H
```

**Comment by SPRESENSE** [comment]
@pkarashchenko 
I have fixed all about your comments. Could you please review them?

**Pull request merged**